conans/cli/command.py

```diff
@@ -0,0 +1,282 @@
+        if getattr(namespace, self.dest) is not None and self.default is None:
+            msg = '{o} can only be specified once'.format(o=option_string)
+            raise argparse.ArgumentError(None, msg)
+        setattr(namespace, self.dest, values)
+
+
+class SmartFormatter(argparse.HelpFormatter):
+
+    def _fill_text(self, text, width, indent):
+        text = textwrap.dedent(text)
+        return ''.join(indent + line for line in text.splitlines(True))
+
+
+class Command(object):
+    """A single command of the conan application, with all the first level commands. Manages the
+    parsing of parameters and delegates functionality to the conan python api. It can also show the
+    help of the tool.
+    """
+    def __init__(self, conan_api):
+        assert isinstance(conan_api, Conan)
+        self._conan = conan_api
+        self._out = conan_api.out
+
+    def help(self, *args):
+        """
+        Shows help for a specific command.
+        """
+        parser = argparse.ArgumentParser(description=self.help.__doc__, prog="conan help",
+                                         formatter_class=SmartFormatter)
+        parser.add_argument("command", help='command', nargs="?")
+        args = parser.parse_args(*args)
+        if not args.command:
+            self._show_help()
+            return
+        try:
+            commands = self._commands()
+            method = commands[args.command]
+            method(["--help"])
+        except KeyError:
+            raise ConanException("Unknown command '%s'" % args.command)
+
+    # conan v2 search:
+    # conan search "*" will search in the default remote
+    # to search in the local cache: conan search "*" --cache explicitly
+
+    def search(self, *args):
+        """
+        Searches for package recipes whose name contain <query> in a remote or in the local cache
+        """
+        parser = argparse.ArgumentParser(description=self.search.__doc__, prog="conan search",
+                                         formatter_class=SmartFormatter)
+        parser.add_argument('query',
+                            help="Search query to find package recipe reference, e.g., 'boost', 'lib*'")
+
+        exclusive_args = parser.add_mutually_exclusive_group()
+        exclusive_args.add_argument('-r', '--remote', default=None, action=OnceArgument, nargs='?',
+                                    help="Remote to search. Accepts wildcards. To search in all remotes use *")
+        exclusive_args.add_argument('-c', '--cache', action="store_true", help="Search in the local cache")
+        parser.add_argument('-o', '--output', default="cli", action=OnceArgument,
+                            help="Select the output format: json, html,...")
+        args = parser.parse_args(*args)
+
+        try:
+            info = self._conan.search_recipes(args.query, remote_pattern=args.remote,
+                                              local_cache=args.cache)
+        except ConanException as exc:
```

Comment:
This repeated pattern everywhere is ugly. I think we should try to avoid it when possible, and extract the behavior to a common place when not possible.
I would remove it at the moment, and leave a flow without try-except.